### PR TITLE
302: Add custom reader for docker logs

### PR DESCRIPTION
**Pull request by lumimul**
PR for #302

### CODE REVIEW

File: src/main/java/com/amihaiemil/docker/ReadLogString.java [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+            try {
+                Charset charset = this.getCharset(contentType);
+                Reader reader = new InputStreamReader(instream, charset);
+                this.read(buffer, reader);
+            } finally {
+                instream.close();
+            }
+        }
+        return buffer.toString();
+    }
```
Comment (amihaiemil) [inline src/main/java/com/amihaiemil/docker/ReadLogString.java:108]: @lumimul Here is a refactored version of this method:

```java
private String toString(final HttpEntity entity) throws IOException {
    final InputStream instream = entity.getContent();
    final CharArrayBuffer buffer = new CharArrayBuffer(
        this.getCapacity(entity)
    );    
    if (instream != null) {
        try {
            final Reader reader = new InputStreamReader(
                instream,
                this.getCharset(ContentType.get(entity))
            );
            this.read(buffer, reader);
        } finally {
            instream.close();
        }
    }
    return buffer.toString();
}
```

I removed the one-time variables, and made all variables ``final``. What do you think? I think it's cleaner this way.

File: src/main/java/com/amihaiemil/docker/ReadLogString.java [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+
+    /**
+     * Read data from reader and save it in buffer.
+     *
+     * @param buffer Buffer for save message.
+     * @param reader Reader for read message.
+     * @throws IOException if the entity cannot be read
+     */
```
Comment (amihaiemil) [inline src/main/java/com/amihaiemil/docker/ReadLogString.java:116]: @lumimul Can you please add a more detailed JavaDoc? Please specify exactly what this method is doing. It contains a very complicated algorithm that nobody will understand later.

File: src/main/java/com/amihaiemil/docker/ReadLogString.java [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+    }
+
+    /**
+     * Get content length of HttpEntity.
+     *
+     * @param entity HttpEntity for get capacity.
+     * @return Capacity.
+     */
+    private int getCapacity(final HttpEntity entity) {
+        Args.check(entity.getContentLength() <= 2147483647L,
```
Comment (amihaiemil) [inline src/main/java/com/amihaiemil/docker/ReadLogString.java:148]: @lumimul shouldn't it be ``> 2147483647L``? 

Also, what exactly does ``2147483647L`` represent?

File: src/main/java/com/amihaiemil/docker/ReadLogString.java [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+                        buffer.append(stdout, 0, len);
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Get content length of HttpEntity.
+     *
```
Comment (amihaiemil) [inline src/main/java/com/amihaiemil/docker/ReadLogString.java:143]: @lumimul Same here, more details in the JavaDoc, please :)

File: src/main/java/com/amihaiemil/docker/ReadLogString.java (hunk 3 above)
Comment (lumimul) [reply src/main/java/com/amihaiemil/docker/ReadLogString.java:148]: image

2147483647 it is Integer.MAX_VALUE

Comment (lumimul) [reply src/main/java/com/amihaiemil/docker/ReadLogString.java:148]: change it

File: src/main/java/com/amihaiemil/docker/ReadLogString.java (hunk 1 above)
Comment (lumimul) [reply src/main/java/com/amihaiemil/docker/ReadLogString.java:110]: yes, use it)

File: src/main/java/com/amihaiemil/docker/ReadLogString.java (hunk 2 above)
Comment (lumimul) [reply src/main/java/com/amihaiemil/docker/ReadLogString.java:132]: add more detail)

File: src/main/java/com/amihaiemil/docker/ReadLogString.java (hunk 4 above)
Comment (lumimul) [reply src/main/java/com/amihaiemil/docker/ReadLogString.java:161]: add more detail)

**Comment by amihaiemil** [comment]
@lumimul thanks a lot! :)

**Comment by amihaiemil** [comment]
@rultor merge it

**Comment by 0crat** [comment]
Job `gh:amihaiemil/docker-java-api#334` is not assigned, can't get performer

**Pull request merged**